python/strings_udf/strings_udf/__init__.py

```diff
@@ -41,59 +28,67 @@ def _get_appropriate_file(sms, cc):
         return None
 
 
+def maybe_patch_numba_linker(driver_version):
+    # Numba thinks cubinlinker is only needed if the driver is older than the ctk
+    # but when strings_udf is present, it might also need to patch because the PTX
+    # file strings_udf relies on may be newer than the driver as well
+    if driver_version < STRINGS_UDF_PTX_VERSION:
+        logger.debug(
+            "Driver version %s.%s needs patching due to strings_udf"
+            % driver_version
+        )
+        if _numba_version_ok:
+            logger.debug("Patching Numba Linker")
+            Linker.new = new_patched_linker
+        else:
+            logger.debug("Cannot patch Numba Linker - unsupported version")
+
+
+def get_ptx_file():
+    dev = cuda.get_current_device()
+
+    # Load the highest compute capability file available that is less than
+    # the current device's.
+    cc = int("".join(str(x) for x in dev.compute_capability))
+    files = glob.glob(os.path.join(os.path.dirname(__file__), "shim_*.ptx"))
+    if len(files) == 0:
+        raise RuntimeError(
+            "This strings_udf installation is missing the necessary PTX "
+            "files. Please file an issue reporting this error and how you "
+            "installed cudf and strings_udf."
+        )
+
+    regular_sms = []
+
+    for f in files:
+        file_name = os.path.basename(f)
+        sm_number = file_name.rstrip(".ptx").lstrip("shim_")
+        if sm_number.endswith("a"):
+            processed_sm_number = int(sm_number.rstrip("a"))
+            if processed_sm_number == cc:
+                return f
+        else:
+            regular_sms.append((int(sm_number), f))
+
+    regular_result = None
+
+    if regular_sms:
+        regular_result = _get_appropriate_file(regular_sms, cc)
+
+    if regular_result is None:
+        raise RuntimeError(
+            "This strings_udf installation is missing the necessary PTX "
+            f"files that are <={cc}."
+        )
+    else:
+        return regular_result[1]
+
+
 # adapted from PTXCompiler
 cp = subprocess.run([sys.executable, "-c", CMD], capture_output=True)
+# must have a driver to proceed
```

Comment:
Sorry, I should have been a bit clearer here.

If we're using cudf in a distributed setup with UCX, this `fork()` call is unsafe (can bring down the infiniband network stack) if it is made after low-level network bringup. Yes, everything is terrible. As a consequence, we need a mechanism to specify the output of `cp` (and hence the driver version) without running this command. In ptxcompiler we did this via an environment variable.

So something like (untested!):

```python
try:
    driver_version = tuple(map(int, os.environ["STRINGS_UDF_CUDA_DRIVER_VERSION"].split(".")))
except KeyError:
    cp = subprocess.run([sys.executable, "-c", CMD], capture_output=True)
    if cp.return_code == 0:
        driver_version = tuple(map(int, cp.stdout.strip().split())[:2]
    else:
        pass
        # TODO should we raise an error here?

...
maybe_patch_number_linker(driver_version)
ptxpath = get_ptx_file() # TODO see comment on context initialisation
```